internal/storage/bucket_handle.go

```diff
@@ -430,6 +435,36 @@ func (b *bucketHandle) ComposeObjects(ctx context.Context, req *gcs.ComposeObjec
 	return
 }
 
+// Fetch bucket type from GetStorageLayout and set it. e.g. Hierarchical or NonHierarchical
+func (b *bucketHandle) FetchAndSetBucketType() {
+	if b.controlClient == nil {
+		b.bucketType = gcs.NonHierarchical
+		return
+	}
+
+	var callOptions []gax.CallOption
+	stoargeLayout, err := b.controlClient.GetStorageLayout(context.Background(), &controlpb.GetStorageLayoutRequest{
+		Name:      "projects/_/buckets/" + b.bucketName + "/storageLayout",
+		Prefix:    "",
+		RequestId: "",
+	}, callOptions...)
+
+	// In case bucket does not exist, set type unknown instead of panic.
+	if err != nil {
+		b.bucketType = gcs.Unknown
+		log.Printf("GetStorageLayout: %v", err)
+		return
+	}
+
+	hierarchicalNamespace := stoargeLayout.GetHierarchicalNamespace()
+	if hierarchicalNamespace != nil && hierarchicalNamespace.Enabled {
+		b.bucketType = gcs.Hierarchical
+		return
+	}
+
+	b.bucketType = gcs.Unknown
```

Comment:
in case hierarchicalNamespace is nil or hierarchicalNamespace is not Enabled, bucket type should be NonHierarchical not unknown